### PR TITLE
lilypond 2.22.1 (new formula)

### PR DESCRIPTION
**Pull request by jeffersonfelixdev**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
This is a pull request to include lilypond 2.22.1 as a formula. The lilypond cask is a 32bit version and it does not work on macOS Catalina or higher. This formula was successfully developed in macOS Big Sur 11.5.2.

There are some considerations:

* the replace in configure options is mandatory to solve conflict with other formula (emacs) and there is no flag to change it;
* guile 3 is not compatible with lilypond build;
* `brew audit --new lilypond` fails because the macOS provided `bison` and `texinfo` are older than the minimum version required to compile lilypond.

Jefferson Felix

### CODE REVIEW

File: Formula/lilypond.rb
Comment (carlocab) [inline Formula/lilypond.rb:41]: We should use the `texinfo` formula.

File: Formula/lilypond.rb
Comment (carlocab) [inline Formula/lilypond.rb:46]: What are these build dependencies? These should be built from source. The only exception is when Lilypond needs itself to bootstrap.

File: Formula/lilypond.rb
Comment (carlocab) [inline Formula/lilypond.rb:31]: This should use the `bison` formula.

File: Formula/lilypond.rb
Comment (carlocab) [inline Formula/lilypond.rb:36]: We should use the upstream project's source for this.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:41]: If I add texinfo formula:
`Dependency 'texinfo' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.`

I need texinfo 6.1 or higher, and the one provided in macOS is older.
Is there a different way to do this?

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:46]: There is some binary files provided by MacTex or TexLive, used only for build.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:31]: same as above (I cannot use bison provided by macOS)

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:46]: I don't know exactly how it works, but lilypond needs bibtex, latex, mf, metapost and other tools from the TeX package (provided by MacTeX or Tex-Live), maybe to compile the documentation.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:36]: ok I will change url

File: Formula/lilypond.rb
Comment (carlocab) [reply Formula/lilypond.rb:41]:
```ruby
depends_on "texinfo"
```

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:31]: Tried to add `uses_from_macs` but got a lot of errors during compilation.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:41]: It does not work in CI:
```
lilypond:
  * Dependency 'bison' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.
  * Dependency 'texinfo' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.
```

Comment (jsfelix) [reply Formula/lilypond.rb:41]: And `uses_from_macs` crashes compiling

Comment (jsfelix) [reply Formula/lilypond.rb:41]: This occours when I add `uses_from_macos`:
```
ERROR: Please install required programs:  bison >= 2.4.1 (installed: 2.3) makeinfo >= 6.1 (installed: 4.8)
```

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:46]: According to the docs (https://lilypond.org/doc/v2.23/Documentation/contributor/requirements-for-compiling-lilypond#other), the following TeX dependencies are required to compile Lilypond:
* MetaFont: The MetaFont binary (usually called mf-nowin, mf, mfw, or mfont) and its support files are normally packaged along with TeX. Most GNU/Linux and other free software distributions already provide packages for TeX Live, see above. TeX Live can can also be installed separately; it contains stand-alone binaries for most platforms.
* MetaPost: The mpost binary is also usually packaged with TeX.
* TeX Gyre font packages in OTF format. Also see Fonts

File: Formula/lilypond.rb
Comment (carlocab) [reply Formula/lilypond.rb:41]: > It does not work in CI:
> 
> ```
> lilypond:
>   * Dependency 'bison' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.
>   * Dependency 'texinfo' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.
> ```

This is just a reminder to use the one that comes with macOS if we can. We can't, so we should do `depends_on "texinfo"` instead of building a throw-away `texinfo` in the formula.

Comment (jsfelix) [reply Formula/lilypond.rb:41]: OK, but `brew audit --new lilypond` fails.

Comment (jsfelix) [reply Formula/lilypond.rb]: ok, done. But i think the CI will fail because `brew audit --new lilypond` fails.

File: Formula/lilypond.rb
Comment (nwhetsell) [reply Formula/lilypond.rb:38]: Since this formula seems to be partially based on

https://github.com/nwhetsell/homebrew-lilypond/blob/63f3be3914d54d19fc86efa234a56019f5727a86/Formula/lilypond.rb#L30-L127

I thought I should note that, over there, I’m handling the dependency on TeX Live as a [requirement on MacTeX](https://github.com/nwhetsell/homebrew-lilypond/blob/63f3be3914d54d19fc86efa234a56019f5727a86/Library/MacTeXRequirement.rb), although I don’t believe this would be suitable for homebrew-core. It seems like https://github.com/Homebrew/homebrew-core/pull/83738 would need to be merged before LilyPond can be included in homebrew-core.

Comment (nwhetsell) [reply Formula/lilypond.rb:38]: I should also note that according to LilyPond’s documentation, using Guile v2 with LilyPond is not officially supported:

https://lilypond.org/doc/v2.22/Documentation/contributor/requirements-for-running-lilypond

Comment (SMillerDev) [reply Formula/lilypond.rb:38]: You should make formula for these dependencies if you need them.

Comment (jsfelix) [reply Formula/lilypond.rb:38]: > Since this formula seems to be partially based on
> 
> https://github.com/nwhetsell/homebrew-lilypond/blob/63f3be3914d54d19fc86efa234a56019f5727a86/Formula/lilypond.rb#L30-L127
> 
> I thought I should note that, over there, I’m handling the dependency on TeX Live as a [requirement of MacTeX](https://github.com/nwhetsell/homebrew-lilypond/blob/63f3be3914d54d19fc86efa234a56019f5727a86/Library/MacTeXRequirement.rb), although I don’t believe this would be suitable for homebrew-core. It seems like #83738 would need to be merged before LilyPond can be included in homebrew-core.

#83738 is linux-only, and this formula is mac-only.

Comment (jsfelix) [reply Formula/lilypond.rb:38]: > You should make formula for these dependencies if you need them.

Is it possible to build these dependencies directly in this formula? Tex Live is very big and already have a cask (mactex and mactex-no-gui)

Comment (nwhetsell) [reply Formula/lilypond.rb:38]: @jsfelix:

> #83738 is linux-only, and this formula is mac-only.

Not necessarily:

* https://github.com/Homebrew/homebrew-core/pull/83738#issuecomment-906190278
* https://github.com/Homebrew/homebrew-core/pull/83738#issuecomment-906195943
* https://github.com/Homebrew/homebrew-core/pull/83738#issuecomment-917685110

Comment (jsfelix) [reply Formula/lilypond.rb:38]: Ok, so I can wait for #83738 and use texlive formula as a dependency.

Comment (jsfelix) [reply Formula/lilypond.rb:38]: > I should also note that according to LilyPond’s documentation, using Guile v2 with LilyPond is not officially supported:
> 
> https://lilypond.org/doc/v2.22/Documentation/contributor/requirements-for-running-lilypond

I am use lilypond built with guile 2 for many years in macOS and I never had problems. Maybe we can keep guile2, or build guile 1.8 inside, because it is a very old and deprecated version.

**Comment by jsfelix** [comment]
I just set up some changes in formula for building and distributing guile 1.8.8 with Lilypond and add `conflicts_with guile, guile@2`, because guile@1 is old and obsolete, and lilypond doesn't look like it will support new versions of guile in the near future.

@carlocab and @nwhetsell, do you agree with this? 

I will push the final version after #83738 merges.

File: Formula/lilypond.rb
Comment (SMillerDev) [inline Formula/lilypond.rb:41]: Are we actually supposed to run this? Usually it's only configure unless upstreams version is broken somehow.

Comment (jsfelix) [reply Formula/lilypond.rb:41]: I removed this line, and consequently the autoconf dependency, and everything still working fine.

**Comment by nwhetsell** [comment]
@jsfelix I’m not a maintainer, just a contributor. I can’t move this one way or the other.

That said, unless I’m missing something:

* This PR hasn’t changed since Sep 17; your changes related to Guile v1 don’t appear here at the moment.

* There are two options for distributing Guile v1 with LilyPond. You could create a `guile@1` formula (like [this](https://github.com/nwhetsell/homebrew-lilypond/blob/304562f7e2262f93abb97dbbad5f7ec8fd13fed1/Formula/guile%401.rb)), but I think it’s unlikely that such a formula would meet the [criteria for versioned formulae](https://docs.brew.sh/Versions) in homebrew-core (in particular the requirement that versioned formulae be used by a large number of people—Guile v1 isn’t and won’t be). The only other possibility that I can think of is to install Guile v1 in LilyPond’s libexec folder before installing LilyPond itself. Either way (and assuming the latter option works at all and is acceptable for homebrew-core), you’ll need a shim script to tell LilyPond where to find Guile v1 (like the one I’m writing [here](https://github.com/nwhetsell/homebrew-lilypond/blob/304562f7e2262f93abb97dbbad5f7ec8fd13fed1/Formula/lilypond.rb#L106-L111), which is based on a script included in [LilyPond’s app bundle](https://lilypond.org/macos-x.html)). Consequently, LilyPond shouldn’t conflict with `guile@2` or `guile`, and a `conflicts_with` statement shouldn’t be necessary.

File: Formula/lilypond.rb
Comment (carlocab) [inline Formula/lilypond.rb:40]: What was wrong with using `guile@2` as in the earlier version of your PR?

In any case, even if we absolutely needed `guile@1`, this should be installed into `libexec` as @nwhetsell suggests.

**Comment by jsfelix** [comment]
@nwhetsell, thanks for your support.

I pushed my new version of the formula, which doesn't use `guile@2`, and built guile@1 along with lilypond, inside the `prefix` folder. For now it should not work as we are waiting for texlive merge.

So I modified the test to include some scheme language, to test lilypond and guile 1.8.8, and I think that's OK.

I don't know if this solution complies with the homebrew-core rules, so the maintainers will be able to comment on that.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:40]: Lilypond needs guile@1 to run and compile: 
https://lilypond.org/doc/v2.23/Documentation/contributor/requirements-for-running-lilypond

> Guile
> Use version 1.8.8. It is possible to build LilyPond with version 2.x of Guile and most features will work, but this is not officially supported (yet). In particular, performance and especially the startup time will be worse.

I'm not very familiar with use of `libexec`, but I will try to install guile on it, and make lilypond to find guile.

Comment (jsfelix) [reply Formula/lilypond.rb:37]: Now guile@1 is in the `libexec` folder.

File: Formula/lilypond.rb
Comment (jsfelix) [inline Formula/lilypond.rb:69]: We will probably change this path depending on what will be decided about using `libexec` in #83738

File: Formula/lilypond.rb
Comment (carlocab) [reply Formula/lilypond.rb:37]: Didn't you have an earlier working version of this formula that used `guile@2`?

Comment (jsfelix) [reply Formula/lilypond.rb:37]: Yes, but lilypond doesn't officially support `guile@2`, although it probably works (https://lilypond.org/doc/v2.22/Documentation/contributor/requirements-for-running-lilypond). I don't know exactly which feature doesn't work, but I've noticed that lilypond works faster with `guile@1`.

Comment (carlocab) [reply Formula/lilypond.rb:37]: We should use `guile@2` unless we know that something's actually broken: https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

Comment (jsfelix) [reply Formula/lilypond.rb]: Restored to guile@2. The new test cover guile and passed without errors.

File: Formula/lilypond.rb
Comment (samford) [inline Formula/lilypond.rb:7]:
```suggestion

  livecheck do
    url "https://lilypond.org/source.html"
    regex(/href=.*?lilypond[._-]v?(\d+(?:\.\d+)+)\.t/i)
  end

```

`brew livecheck` unable to identify versions by default, so this needs a `livecheck` block to work. The suggested `livecheck` block above checks the first-party download page for the source tarball, which links to the `stable` archive.

File: Formula/lilypond.rb
Comment (nwhetsell) [inline Formula/lilypond.rb:50]: `guile@2` doesn’t install a libexec folder. `Formula["guile@2"].opt_libexec/"lib"` (which doesn’t exist) should be `Formula["guile@2"].lib`.

Comment (jsfelix) [reply Formula/lilypond.rb:50]: Oh, yes, I forgot to change this line, but I'm not sure if this is still necessary as it worked even with the wrong path.

**Comment by jsfelix** [comment]
According to the developers of Lilypond (https://lists.gnu.org/archive/html/lilypond-devel/2021-10/msg00007.html), it is highly recommended to use Guile 1.8.8. They even mention that the major Linux distributions that ship Lilypond have also agreed to use Guile 1.8.8, until Lilypond officially supports Guile 2.

As they haven't reported any known issues, I'm inclined to keep Guile 2 in favor of homebrew rules.

**Comment by carlocab** [comment]
> As they haven't reported any known issues, I'm inclined to keep Guile 2 in favor of homebrew rules.

Yep, let's stick with Guile 2 unless we know of any issues. If we do discover any we can either try to patch it, or switch to the recommended Guile version.

**Comment by jsfelix** [comment]
I pushed the final version of this formula, and now I think everything is OK.

File: Formula/lilypond.rb
Comment (carlocab) [inline Formula/lilypond.rb:34]: Is there no `configure` flag we can use to customise this?

File: Formula/lilypond.rb
Comment (carlocab) [inline Formula/lilypond.rb:51]: Use `write_env_script`. Just `grep` the core repo for examples.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:34]: I searched for lilypond source and I cannot find a flag to configure this directory.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb]: I just updated the formula to use `write_env_script`.

Comment (jsfelix) [reply Formula/lilypond.rb]: After some testing, I found a solution by changing the `datadir` option in configure. I'm making the changes and will send you the solution in a moment.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:34]: After some testing, I found a solution by changing the `datadir` option in configure. I'm making the changes and will send you the solution in a moment.

Comment (jsfelix) [reply Formula/lilypond.rb]: I uploaded a change, but apparently it didn't work. There definitely doesn't seem to be a simple solution, as `make install` is creating links in `/usr/local/share`, even if you enter another `prefix`. I'll study this further and maybe get in touch with the developers for help.

Comment (jsfelix) [reply Formula/lilypond.rb]: OK, I think I understand now. I'm not very familiar with `emacs`, but as I've seen in other formulas, like `libidn` and `clisp`, elisp files are linked to `/usr/local/share/emacs/site-lisp`. The correct placement that the homebrew rules suggest is to place these links in `/usr/local/share/emacs/site-lisp/{formula-name}` to avoid conflicts with other formulas. But lilypond install creates these links directly on the `site-lisp`. I'll find a solution to put it in the right place without making substitutions to `config.make.in`.

File: Formula/lilypond.rb
Comment (carlocab) [inline Formula/lilypond.rb:41]: Setting `--datadir` didn't work, unfortunately.
```
lilypond:
  * Emacs Lisp files were linked directly to "/opt/homebrew/share/emacs/site-lisp".
    This may cause conflicts with other packages.
    They should instead be installed into:
      /opt/homebrew/opt/lilypond/share/emacs/site-lisp/lilypond
    The offending files are:
      /opt/homebrew/opt/lilypond/share/emacs/site-lisp/lilypond-what-beat.el
            /opt/homebrew/opt/lilypond/share/emacs/site-lisp/lilypond-words.el
            /opt/homebrew/opt/lilypond/share/emacs/site-lisp/lilypond-song.el
            /opt/homebrew/opt/lilypond/share/emacs/site-lisp/lilypond-font-lock.el
            /opt/homebrew/opt/lilypond/share/emacs/site-lisp/lilypond-init.el
            /opt/homebrew/opt/lilypond/share/emacs/site-lisp/lilypond-indent.el
            /opt/homebrew/opt/lilypond/share/emacs/site-lisp/lilypond-mode.el
```

But what we can do is this:

```suggestion

    elisp.install share.glob("emacs/site-lisp/*.el")
```
which is less fragile than the `inreplace`.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb]: tried: 
```
"--with-lispdir=#{elisp}"
```
without success: 
```
configure: WARNING: unrecognized options: --with-lispdir
```

Comment (carlocab) [reply Formula/lilypond.rb]: You can keep the `--datadir` option and apply the suggestion I made below.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb:40]: oops, I only see your message now. Please verify if my solution I just pushed works too.

Comment (carlocab) [reply Formula/lilypond.rb:40]: I think we want to set `--datadir` anyway, just in case that reference is compiled into the software and used somewhere. I also slightly prefer `foo.glob` to using `Dir[]`, but that's just personal preference.

Comment (jsfelix) [reply Formula/lilypond.rb:40]: ok, I will make these changes.

**Comment by jsfelix** [comment]
I just pushed the changes suggested by @carlocab and now it will be ready to merge.

**Comment by jsfelix** [comment]
What happened to the mojave bottle?

**Review by carlocab — DISMISSED** [review]
Thanks, @jsfelix!

File: Formula/lilypond.rb
Comment (nwhetsell) [inline Formula/lilypond.rb:46]: Should this be

```ruby
(bin/"lilypond").write_env_script libexec/"lilypond",
```

(not `opt_libexec/"lilypond"`), or am I missing something?

Comment (carlocab) [reply Formula/lilypond.rb]: They're almost always the same, but `libexec` is a better choice for some edge cases. Let's do that.

**Review by carlocab — APPROVED** [review]
Will merge this when CI finishes. Ping me if I forget to.

File: Formula/lilypond.rb
Comment (jsfelix) [reply Formula/lilypond.rb]: In previous version, we used `write` statement instead of `write_env_script`, but the script is the same.

**Pull request closed without merging**
